test/types/instance.test-d.ts

```diff
@@ -6,6 +6,36 @@ import { HookHandlerDoneFunction } from '../../types/hooks'
 
 const server = fastify()
 
+server.decorate('nonexistent', () => {})
+server.decorateRequest('nonexistent', () => {})
+server.decorateReply('nonexistent', () => {})
+
+declare module '../../fastify' {
+  interface FastifyInstance {
+    functionWithTypeDefinition: (foo: string, bar: number) => Promise<boolean>
+  }
+  interface FastifyRequest {
+    numberWithTypeDefinition: number
+  }
+  interface FastifyReply {
+    stringWithTypeDefinition: 'foo' | 'bar'
+  }
+}
+expectError(server.decorate('functionWithTypeDefinition', (foo: any, bar: any) => {})) // error because invalid return type
+expectError(server.decorate('functionWithTypeDefinition', (foo: any, bar: any) => true)) // error because doesn't return a promise
+expectError(server.decorate('functionWithTypeDefinition', async (foo: any, bar: any, qwe: any) => true)) // error because too many args
+server.decorate('functionWithTypeDefinition', async (foo, bar) => {
+  expectType<string>(foo)
+  expectType<number>(bar)
+  return true
+})
+
+expectError(server.decorateRequest('numberWithTypeDefinition', 'not a number')) // error because invalid type
```

Comment:
I'd also add a test to guarantee that `decorateRequest`/`decorateReply` works as expected if the params are valid.

Reply:
Something in addition to lines 34 and 37?

Reply:
That's ok!